GCTC-NTGC/gc-digital-talent
# Fancy filter

Resolves #2641

Prototype: https://www.figma.com/proto/XS4Ag6GWcgdq2dBlLzBkay/Admin-v3-(Profile-release)?node-id=465%3A8698&scaling=min-zoom&page-id=0%3A1&starting-point-node-id=410%3A1984&show-proto-sidebar=1&hide-ui=1

## To Do
- [x] wire up clear button
- [x] ensure all strings localized
- [x] figure out localized select options
- [x] ~~import alternative grouped multiselectfield (was building demo outside repo)~~
- [x] wire up to search filter
- [x] jest tests
  - waiting for review first?
- [x] fix test that's timing out
- vachan feedback
  - [x] a6d3d43 a496b66 fix CSS issue 
  - [x] embed in User.stories.jsx "User Table" stories
  - [x] 0ba68ca fixed storybook brokenness
- peter feedback
  - [x] 3ccc00d7b create constants for `eductionRequirements`
  - [x] 1d79ba29a use `notEmpty` instead of new `notNullOrUndefined`
  - [x] 4a4d34cd8 improve docs of latency simulation in `MockGraphqlDecorator`
  - [x] e21acf01b avoid using `<FormattedMessage />` component
  - [x] add jest test(s) for `SelectFieldV2` component
  - [x] add jest test(s) for `MultiSelectFieldV2` component
  - [x] add jest test(s) for `UserTableFilterDialog` component (test filter object)
  - [x] add jest test(s) for `useFilterOptions` hook
- [x] split SelectFieldV2 and MultiSelectFieldV2 components/tests/stories into own PR https://github.com/GCTC-NTGC/gc-digital-talent/pull/3510
- tristan feedback
  - [x] 8976c920a move useLocale to common dir https://github.com/GCTC-NTGC/gc-digital-talent/pull/3249/files#r924925000

## Manual QA

1. Visit "SearchFilter > Default" component in storybook
2. Click the **dialog's "x"** in the corner and re-open.
3. With empty filters...
    - **Submit** and note **empty arrays** for each field's data in Action output.
5. **Add** some filter options to Select and MultiSelect fields. Clear each with **field's "x"** button.
    - **Submit** and note **empty arrays** for each field's data in Action output. Re-open.
6. **Add** some filter options to Select and MultiSelect fields, then close vis dialog's dismiss "x". Re-open and note **empty filters** (**no persistence**).
    - **Submit** and note **empty arrays** for each field's data in Action output. Re-open.
8. Add some filter options to Select and MultiSelect fields, then click **"Clear"** button. Fields should all empty.
    - **Submit** and note **empty arrays** for each field's data in Action output. Re-open.
9. **Add** some filter options to Select and MultiSelect fields.
    - **Submit** and note **data arrays** (for both Select and MultiSelect) for field's data in Action output.
    - **Re-open** and note that filter values **have persisted**.
10. With persisted filter state, **click "Clear"** and note that all filters clear.
    - Click **dialog's "x"** to dismiss modal.
    - Re-open and note that previously submitted filter values are back (clearing **did not persist**).
11. With persisted filter state, **click "Clear"** and then **"Submit"**.
    - Re-open and note empty fields (clearing **did persist**).
12. ~~Localization: Change locale and note that labels all change to French (WIP)~~ (can't see this until translated)
13. ~~Localization: Change locale and note that select options all change to French (WIP)~~ (can't see this until translated)
